### PR TITLE
CGC: Incremental Run - Default behavior

### LINKED ISSUE #5783: GC: Incremental - Read first run ID when using the incremental flag
To implement incremental GC, implement the following default behavior:
If the `lakefs.gc.incremental` flag was introduced with the value `true` (`spark.hadoop.lakefs.gc.incremental=true`), then the GC will choose the latest run ID from the run id list and will send it to the lakeFS server for further processing. If no run ID was found, it will abort.

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
### Linked Issue

Closes #5783

---

## Change Description

This PR includes the default behavior of enabled incremental CGC run.
The following is the behavior implemented:
1. Once the `lakefs.gc.incremental` flag is turned, CGC will try to fetch the **latest** completed GC run.
2. If there're no previous run IDs it will fail, if there is at least one, it will use it as an input for lakeFS's `prepareCommits` API.

This PR also includes some refactoring. The refactor isn't 100% complete since some functionality that is currently in the `BulkRemover` should be moved to the `StorageClient`s. This will happen on a later PR.

## How was this tested?

Manually covering scenarios over `S3` and `Azure`:
1. Full GC.
2. Incremental - one Run ID available.
3. Incremental - two Run IDs available.
4. Incremental - no Run ID available.

## Important!

This is not the last PR of this feature. The next one will make use of an `iteration` parameter that will let the user choose how many iterations of Run IDs should we go back

### Even more important!

Tests will be added.

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 1]
```diff
@@ -543,6 +544,41 @@ object GarbageCollector {
...
+      storageNS: String,
+      configMapper: ConfigMapper
+  ): String = {
+    if (!incrementalRun) {
+      return ""
+    }
+    var previousRunID = ""
+    val runIDsPath = new Path(s"${storageNS.stripSuffix("/")}/_lakefs/retention/gc/run_ids/")
+    val runIDsFS = runIDsPath.getFileSystem(configMapper.configuration)
+    val run_ids = runIDsFS.listFiles(runIDsPath, false)
```
Comment (johnnyaug) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:558]:
```suggestion
    val runIDs = runIDsFS.listFiles(runIDsPath, false)
```

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 2]
```diff
@@ -543,6 +544,41 @@ object GarbageCollector {
...
+    } else {
+      runIDsFS.close()
+      throw RunIDException("No previous run ID")
+    }
+    runIDsFS.close()
+    previousRunID
+  }
+
+  private def logRunID(storageNS: String, runID: String, configMapper: ConfigMapper): Unit = {
+    val runIDPath = new Path(s"${storageNS.stripSuffix("/")}/_lakefs/retention/gc/run_ids/$runID")
```
Comment (johnnyaug) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:571]: The base path can be a constants

**Review by johnnyaug — APPROVED** [review]
LGTM

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala]: done

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 2 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:571]: doing in next PR

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 3]
```diff
@@ -543,6 +544,41 @@ object GarbageCollector {
...
+    var previousRunID = ""
+    val runIDsPath = new Path(s"${storageNS.stripSuffix("/")}/_lakefs/retention/gc/run_ids/")
+    val runIDsFS = runIDsPath.getFileSystem(configMapper.configuration)
+    val runIDs = runIDsFS.listFiles(runIDsPath, false)
+    if (runIDs.hasNext) {
+      previousRunID = runIDs.next().getPath.getName
+      println(s"----------------------- Using previous RUN ID: $previousRunID")
+    } else {
+      runIDsFS.close()
+      throw RunIDException("No previous run ID")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:564]: As I suggest in #5867, here too I prefer simply to run non-incrementally.  Think of the first GC run on a repo: no reason to fail it.  (Especially true in lakeFS.Cloud or other managed GC situations).

File: clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala [hunk 4]
```diff
@@ -1,95 +1,47 @@
...
+      extends StorageClient
+      with Serializable {
+    private val storageNSURI: URI = new URI(storageNamespace)
+    private val storageAccountUrl: String =
+      StorageUtils.AzureBlob.uriToStorageAccountUrl(storageNSURI)
+    private val storageAccountName: String =
+      StorageUtils.AzureBlob.uriToStorageAccountName(storageNSURI)
+    @transient private lazy val blobServiceClient: BlobServiceClient =
+      getBlobServiceClient(storageAccountUrl, storageAccountName, config)
+    @transient lazy val blobBatchClient: BlobBatchClient = new BlobBatchClientBuilder(
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala:37]: This one is not private?

File: clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala [hunk 5]
```diff
@@ -125,14 +77,18 @@ object StorageClients {
...
+  def apply(
+      storageType: String,
+      configMapper: ConfigMapper,
+      storageNamespace: String,
+      region: String
+  ): StorageClient = {
     storageType match {
       case StorageTypeS3 =>
-        new S3(configMapper)
+        new S3(storageNamespace, region, S3.S3NumRetries, configMapper)
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala:88]: Prefer to have configMapper either always first or always last.  Here it seems opposite to l. 90.

**Review by arielshaqed — APPROVED** [review]
Thanks!

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 3 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:564]: lets discuss that on #5867

**Pull request merged**